Radix history unit tests

maximus19910112 opened the pull request:
## Overview
Creating a test specification for RadixTree implementation

### Notes
For writing tests, I used examples from module [HistoryActionTests.scala](https://github.com/maximus19910112/rchain/blob/c2535eb8ce3d65cc417a1e56af9ae70e1af647c9/rspace/src/test/scala/coop/rchain/rspace/history/HistoryActionTests.scala).
I appended unit-tests, that validate work of main procedures of module [RadixTree.scala](https://github.com/maximus19910112/rchain/blob/c2535eb8ce3d65cc417a1e56af9ae70e1af647c9/rspace/src/main/scala/coop/rchain/rspace/history/RadixTree.scala):
1. RadixTreeImpl.update:
    -    Creating tree with one leaf
    -    Creating node with two leafs
    -    Creating tree consisting of 1 nodePtr and 2 leafs
    -    Updating leaf data
    -    Assertion check: update does not allow to enter keys with different lengths
    -    Assertion check: update does not allow to radix key is smaller than NodePtr key
    -    Check: deleting non - existent data should return none
2. RadixTreeImpl.delete:
    -    Deleting non-existent data returns none
    -    Destroy tree that consists of 1 node should destroy tree
    -    Delete leaf from node with 2 leafs leaves node with 1 leaf
    -    Delete leaf from tree with 1 nodePtr and 2 leafs leaves tree with 1 leaf(NodePtr is also destroyed)
3. RadixTreeImpl.read:
    -   Reading existing data should return data(in case data don't exist it should return none)
4. RadixTreeImpl.makeActions:
    -   Function should build tree with many leafs and nodePtr's correctly
    -   Assertion check: function should detect collision in KVDB
    -   Function should not create artefacts (no creating excess nodes while it works)
5. RadixTreeImpl.saveNode: 
    -   Function should save selected node into store
6. Functions RadixTree.Codecs.encode and RadixTree.Codecs.decode should work correctly. Also appended check for decoding wrong serialized data
7. Appended exception handling in function RadixTree.Codecs.decode
8. Refactoring of module RadixTreeSpec according to @tgrospic @stanislavlyalin and @hilltracer code review

### Please make sure that this PR:
- [x] is at most 200 lines of code (excluding tests),
- [x] meets [RChain development coding standards](https://rchain.atlassian.net/wiki/spaces/DOC/pages/28082177/Coding+Standards),
- [x] includes tests for all added features,
- [x] has a reviewer assigned,
- [x] has [all commits signed](https://rchain.atlassian.net/wiki/spaces/DOC/pages/498630673/How+to+sign+commits+to+rchain+rchain).

- `bors r+` runs integration tests and merges the PR (if it's approved),
- `bors try` runs integration tests for the PR,
- `bors delegate+` enables non-maintainer PR authors to run the above.

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,628 @@
+    (impl, _) =>
+      val keys = Vector[ByteVector](
+        createBV("0001122013"),
+        createBV("0001122225")
```
Please use this data set: 
```
val dataSet = List( 
   radixKV("1122334455", "01"), 
   radixKV("112233AABB", "02"))
```

hilltracer replied:
where
```scala
case class radixKV (rKey: ByteVector, rValue: ByteVector)

object radixKV {
  def apply(strKey: String, strValue: String): radixKV = {
    new radixKV(createBV(strKey), createBV32(strValue))
  }
}
```

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,628 @@
+class RadixTreeSpec extends FlatSpec with Matchers with OptionValues with InMemoryHistoryTestBase {
+  "appending leaf in empty tree" should "create tree with one node" in withImplAndStore {
+    (impl, _) =>
+      for {
```
Please use this data set: `val dataSet = radixKV("1122334455", "01")`

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,628 @@
+                       )
+
+        referenceTree2 = Vector(
+          "TREE WITH ONE NODE AND 2 LEAFS: root =>",
```
TREE WITH ONE NODE POINTER AND 2 LEAFS

hilltracer replied:
TREE WITH ONE NODE POINTERS AND 2 LEAFS

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,628 @@
+      val keys = Vector[ByteVector](
+        createBV("00000000"),
+        createBV("34564544")
+      )
```
Please use this data set:
```
val dataSet = List( 
   radixKV("1122334455", "01"), 
   radixKV("AABBCCDDEE", "02"))
```

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,628 @@
+
+  def createBV32(s: String): ByteVector = {
+    val notEmptyPart = createBV(s)
+    val emptyPart    = List.fill(32 - notEmptyPart.size.toInt)(0).map(_.toByte)
```
val emptyPart    = List.fill(32 - notEmptyPart.size.toInt)(0x00.toByte)

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,628 @@
+      ("33", "04"),
+      ("FF0011", "05"),
+      ("FF012222", "06")
+    )
```
I think such better:
```scala
case class radixKV (rKey: ByteVector, rValue: ByteVector)

object radixKV {
  def apply(strKey: String, strValue: String): radixKV = {
    new radixKV(createBV(strKey), createBV32(strValue))
  }
}

val dataSet = List(
  radixKV("111122334455", "01"),
  radixKV("11112233AABB", "02"),
  radixKV("1111AABBCC", "03"),
  radixKV("33", "04"),
  radixKV("FF0011", "05"),
  radixKV("FF012222", "06")
)
```

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,628 @@
+        item1 <- impl.update(
+                  RadixTree.EmptyItem,
+                  createBV("FFFFFFF1"),
+                  createBV32("A")
```
dataSet.rKey,
dataSet.rValue

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,628 @@
+    val firstLeafData = createBV32("0xCB")
+    val newLeafData   = createBV32("0xFF")
+    val leafKey       = createBV("00123456F1")
+    for {
```
initiaKVPair = radixKV("1122334455", "01")
newKVPair = radixKV("1122334455", "FF")

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,628 @@
+                     )
+
+      item2Opt  <- impl.update(item1Opt.get, leafKey, newLeafData)
+      _         <- Sync[Task].delay(byteToInt(leafKey.head))
```
It's a useless operation, please delete it.

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,628 @@
+    (impl, _) =>
+      val leafData    = createBV32("CB")
+      val leafKey     = createBV("00123456F1")
+      val testLeafKey = createBV("0112")
```
initiaKVPair = radixKV("1122334455", "01")
wrongKVPair = radixKV("112233", "02")

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,628 @@
+                  createBV("00"),
+                  createBV32("33")
+                )
+                .attempt
```
Please use this initial data
val initialItem = NodePtr(createBV("11223344"), createBV32("01")),
val wrongKVPair = radixKV("11", "FF")

...  .update(
                  initialItem,
                  wrongKVPair.rKey,
                  wrongKVPair.rValue
                )

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,628 @@
+
+  "deleting non - existent data" should "return none" in withImplAndStore { (impl, _) =>
+    val leafData = createBV32("CC")
+    val leafKey  = createBV("00123456F1")
```
val initialKVPair = radixKV("1122334455", "01")
val nonExistingKey = createBV("FFFFFFFFFF")

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,628 @@
+  "deleting leaf from tree with only one leaf" should "destroy tree" in withImplAndStore {
+    (impl, _) =>
+      val leafData = createBV32("CC")
+      val leafKey  = createBV("00123456F1")
```
val initialKVPair = radixKV("1122334455", "01")

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,628 @@
+      )
+
+      val rootItem1Hash = createBV32("11")
+      val rootItem2Hash = createBV32("AF")
```
val dataSet = List( 
   radixKV("1122334455", "01"), 
   radixKV("AABBCCDDEE", "02"))

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,628 @@
+    (impl, _) =>
+      val keys = Vector[ByteVector](
+        createBV("FA01122013"),
+        createBV("FA01122225")
```
val dataSet = List( 
   radixKV("1122334455", "01"), 
   radixKV("11223344FF", "02"))

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,628 @@
+
+        itemIdx     <- Sync[Task].delay(byteToInt(keys(0).head))
+        deletedItem <- impl.delete(rootNode1(0xFA), keys(0).tail)
+        rootNode2   = rootNode1.updated(itemIdx, deletedItem.get)
```
I don't understand why you did this. This is very wrong place =)
item3Opt <- impl.delete(item2Opt, dataSet(1).rKey)
rootNode2 <- impl.constructNodeFromItem(item3Opt.get)

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,628 @@
+
+  "reading data from existing node" should "return data" in withImplAndStore { (impl, _) =>
+    val itemData = createBV32("CB")
+    val key      = createBV("0123456F1")
```
val initialKVPair = radixKV("1122334455", "01")

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,628 @@
+
+  "reading non - existent data" should "return none" in withImplAndStore { (impl, _) =>
+    val itemData = createBV32("CB")
+    val key      = createBV("0123456F1")
```
val initialKVPair = radixKV("1122334455", "01")

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,628 @@
+      newBuf.put(arr).rewind()
+    }
+
+    val collisionKVPair = (copyBVToBuf(History.emptyRootHash.bytes), ByteVector(0x00))
```
please add 
```scala
val emptyRootHash: Blake2b256Hash = Blake2b256Hash.fromByteVector(hashNode(emptyNode)._1)
```
in RadixTree object and use this value in this place

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,628 @@
+        ("33", "04"),
+        ("FF0011", "05"),
+        ("FF012222", "06")
+      )
```
val dataSet = List(
  radixKV("111122334455", "01"),
  radixKV("11112233AABB", "02"),
  radixKV("1111AABBCC", "03"),
  radixKV("33", "04"),
  radixKV("FF0011", "05"),
  radixKV("FF012222", "06")
)

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,628 @@
+    for {
+      //  Create tree with 6 leafs
+      rootNode2Opt <- impl.makeActions(emptyNode, insertActions)
+      hash         = rootNode2Opt.map(node => impl.saveNode(node))
```
hash = impl.saveNode(rootNode2Opt.get)

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,628 @@
+      (
+        ExportData(_, nodeKVDBKeys, nodeKVDBValues, _, _),
+        _
+      ) = exported1
```
please use this way for parsing: 
nodeKVDBKeys = exported1._1.nodeKeys
nodeKVDBValues = exported1._1.nodeValues

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,628 @@
+          None,
+          0,
+          100,
+          newLocalStore.get1,
```
Instead of creating new storage better to use this expressions:
```scala
x => Sync[Task].delay(localStorage.get(x))
```
where
```scala
localStorage = (nodeKVDBKeys zip nodeKVDBValues).toMap
```

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,628 @@
+      _ = exportData2.leafValues shouldBe referenceLeafValues
+    } yield ()
+  }
+
```
Please add two tests for export: `multipage export test`, and `multipage export test with skip`. 
For an example, you can use these tests: 
https://github.com/rchain/rchain/blob/0360a245e62b642579f61c6ba6003e6aea317c25/rspace/src/test/scala/coop/rchain/rspace/ExportImportTests.scala#L83
https://github.com/rchain/rchain/blob/0360a245e62b642579f61c6ba6003e6aea317c25/rspace/src/test/scala/coop/rchain/rspace/ExportImportTests.scala#L164
This is good task for your export understanding, because we panning to realized new parallel Export algorithm.

hilltracer replied:
please add a test for this exception: 
https://github.com/rchain/rchain/blob/0360a245e62b642579f61c6ba6003e6aea317c25/rspace/src/main/scala/coop/rchain/rspace/history/RadixTree.scala#L543
also please add test for validate how works export with non-Existing tree:
https://github.com/rchain/rchain/blob/0360a245e62b642579f61c6ba6003e6aea317c25/rspace/src/main/scala/coop/rchain/rspace/history/RadixTree.scala#L545

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,628 @@
+  ): List[InsertAction] =
+    tuplesKeyAndHash.map {
+      case (key, data) =>
+        InsertAction(createBV(key).toSeq, Blake2b256Hash.fromByteVector(createBV32(data)))
```
```scala
tuplesKeyAndHash: List[radixKV]
): List[InsertAction] =
    tuplesKeyAndHash.map {x =>
        InsertAction(x.rKey.toSeq, Blake2b256Hash.fromByteVector(x.rValue))
```

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,628 @@
+    val wrongSerialized = createBV(
+      "0102FFFF00000000000000000000000000000010280FFFFFFFFFFFFFFFFFFFFFFFFFF"
+    )
+    try {
+      RadixTree.Codecs.decode(wrongSerialized)
+    } catch {
+      case ex: AssertionError =>
+        ex shouldBe a[AssertionError]
+        ex.getMessage shouldBe s"assertion failed: Error during deserialization: invalid data format"
+
+    }
```
Please use `_.attempt` for intercept error, and delete try and catch.

maximus19910112 replied:
I tried to use this construction: 

    for {
      err <- Sync[Task].delay(RadixTree.Codecs.decode(wrongSerialized)).attempt

      ex = err.left.get
      _  = ex shouldBe a[AssertionError]
      _  = ex.getMessage shouldBe referenceErrorMsg
      _  = println(ex.getMessage)
    } yield ()

but it's not work correctly. I cannot catch exception. I also tried to use this code:

    for {
      err <- RadixTree.Codecs.decode(wrongSerialized.pure[Task].attempt
       ...
    } yield ()

no effect

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,830 @@
+
+import java.nio.ByteBuffer
+import scala.concurrent.duration._
+import scala.util.Random
```
Please, use `code->optimize Imports` in Idea

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,830 @@
+      } yield ()
+  }
+
+  "deleting one leaf from a child node containing two leafs" should "child node and reformat parent node" in withImplAndStore {
```
should "**delete** child node and reformat..

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,830 @@
+    } yield ()
+  }
+
+  "function saveNode" should "put node into store" in withImplAndStore { (impl, inMemoStore) =>
```
Сall of functions saveNode() and then commit() should

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,830 @@
+  }
+
+  // Data for test are given from RadixTree specification
+  "encode and decode" should "give initial node" in {
```
encoding and then decoding a node should give this node

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,830 @@
+      newBuf.put(arr).rewind()
+    }
+
+    val emptyRootHash: Blake2b256Hash = Blake2b256Hash.fromByteVector(hashNode(emptyNode)._1)
```
![image](https://user-images.githubusercontent.com/81892279/162587404-b704539f-49fd-4777-bc4a-0429af171dd9.png)

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,830 @@
+    val emptyRootHash: Blake2b256Hash = Blake2b256Hash.fromByteVector(hashNode(emptyNode)._1)
+    val collisionKVPair               = (copyBVToBuf(emptyRootHash.bytes), ByteVector(0x00))
+    val referenceErrorMessage = s"1 collisions in KVDB (first collision with key = " +
+      s"${History.emptyRootHash.bytes.toHex})."
```
At this moment `History.emptyRootHash` and `RadixTree.emptyRootHash` is the same. But better to use `RadixTree.emptyRootHash` for your tests.

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,830 @@
+        )
+      }
+
+      //  Data exported from created storage must me equal to data from source store
```
must **be**

hilltracer replied:
Please use **1** space after all `//` (e.g., `// Data ex...`)

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,830 @@
+
+        _ = tree1 shouldBe referenceTree1
+        _ = nodesCount1 shouldBe 4
+        _ = tree2 shouldBe referenceTree2
```
please add 
```scala
rootNode2Opt shouldBe Some(RadixTree.emptyNode)
```

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,830 @@
+      _ = {
+        assert(exported2 == exported1, "Error of validation")
+        ()
+      }
```
Please delete this assert(), because already we have ` _ = exported1 shouldBe exported2`

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,830 @@
+      _ = {
+        assert(firstExportData == reconstructExportData, "Error of validation")
+        ()
+      }
```
this expression may be remove, because we have `_ = firstExport shouldBe reconstructExport` in tests.

hilltracer commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,830 @@
+          prevData.leafPrefixes ++ pageData.leafPrefixes,
+          prevData.leafValues ++ pageData.leafValues
+        )
+      parameters match {
+        case ExportParameters(
+            rootHash,
+            store,
+            takeSize,
+            skipSize,
+            withSkip,
+            exportData,
+            lastPrefix
+            ) =>
```
Instead of this match you can give value directly from this class, e.g.: `if (parameters.withSkip) parameters.skipSize else 0`. For economy place you can rename input data `parameters` to e.g `p`: `if (p.withSkip) p.skipSize else 0`. In this case user see that this val it's init data from `p`.

hilltracer reviewed (commented):
I hope this is the last time. I'll try not to hurt you anymore.

hilltracer commented:
Please to do next refactoring: change output makeAction() form: `F[Option[Node]]`, to `F[Option[(Node, ByteVector)]]`. Where second parameter is hash of new rootNode. For this - you should rewrite 
https://github.com/rchain/rchain/blob/4fe2728e4f35951f31784f3fce6dc488a3ca564a/rspace/src/main/scala/coop/rchain/rspace/history/RadixTree.scala#L915
to 
```scala
yield if (newCurNode != curNode) (newCurNode, saveNode(newCurNode)).some else none
```
Also please adapt description for this function.
Please rewrite all place where use this function:

hilltracer commented:
Sorry..

hilltracer committed:
Update tree description.

hilltracer commented:
@maximus19910112 I [add](https://github.com/rchain/rchain/pull/3637/commits/71bdbca696b77d6061d99c0b3680e0979a876489)  changingTree() function in RadixTreeImp. Please update your tests: instead `makeActions()` we should to use this new method. If you change it you don't need to call saveNode() and commit after each call changingTree().

tgrospic commented:
> I add changingTree() function in RadixTreeImp

@hilltracer This function can be called `saveAndCommit` to be more clear. Because it's on the tree object it not need to repeat _tree_ in the name.

hilltracer reviewed (approved):
Good work!

maximus19910112 committed:
Refactoring of RadixTree and RadixHistory:
 1) Append to RadixTree module value emptyRootHash
 2) Changed function name: changingTree replaced with saveAndCommit
 3) Little refactoring of emptyRootHash
 4) Refactoring of Codecs.decode method

Original exception supplied to wrapped exception

maximus19910112 committed:
Creating unit - tests

stanislavlyalin commented on rspace/src/main/scala/coop/rchain/rspace/history/RadixTree.scala:
```diff
@@ -201,7 +214,13 @@ object RadixTree {
         }
 
-      decodeItem(0, emptyNode)
+      try {
+        decodeItem(0, emptyNode)
+      } catch {
+        case _: Exception =>
+          assert(assertion = false, "Error during deserialization: invalid data format")
+          emptyNode
+      }
```
This code confuses me, but I don't know how to write it correctly. In the current codebase, I did not find the use of try-catch. Why use assert instead of exception? Maybe consider using an Option, Either, or an Attempt? @tgrospic, what do you think?

tgrospic replied:
If we cannot raise Applicative error (@hilltracer have measured slowdown) we can use suppress warning annotation which will allow the use of `throw` so that original exception can be wrapped.

```scala
@SuppressWarnings(Array("org.wartremover.warts.Throw"))
def decode(bv: ByteVector): Node = {
  ...
  try {
    decodeItem(0, emptyNode)
  } catch {
    case cause: Exception =>
      throw new Exception("Error during deserialization: invalid data format", cause)
  }
}
```

stanislavlyalin commented on rspace/src/main/scala/coop/rchain/rspace/history/instances/RadixHistory.scala:
```diff
@@ -14,7 +14,7 @@ import scodec.bits.ByteVector
   */
 object RadixHistory {
-  val emptyRootHash: Blake2b256Hash = Blake2b256Hash.fromByteVector(hashNode(emptyNode)._1)
+  val emptyRootHash: Blake2b256Hash = RadixTree.emptyRootHash
```
It is just duplicates the same value from `RadixTree`. I think we can eliminate it.

stanislavlyalin commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,790 @@
+        rootNodeAndHash1Opt <- impl.saveAndCommit(RadixTree.emptyNode, referenceInsertActions)
+        (rootNode1, _)      = rootNodeAndHash1Opt.get
+
+        //    Get the tree for compare with reference
```
Different number of spaces in comment lines. We should use single space after slashes. Please check this whole file.

stanislavlyalin commented on rspace/src/test/scala/coop/rchain/rspace/history/RadixTreeSpec.scala:
```diff
@@ -0,0 +1,790 @@
+    } yield result
+  }
+
+  protected def withImplAndStore(
```
This function can be `private`.

stanislavlyalin reviewed (commented):
Good work! Thanks for rebasing commits. I just add some minor issues.

tgrospic commented on rspace/src/main/scala/coop/rchain/rspace/history/RadixTree.scala:
```diff
@@ -201,7 +215,12 @@ object RadixTree {
+        decodeItem(0, emptyNode)
+      } catch {
+        case _: Exception =>
+          throw new Exception("Error during deserialization: invalid data format")
```
Original exception must be supplied to wrapped exception. Please see my example in previous comment.

tgrospic reviewed (approved):
Very good work! Congratulations on your first merged PR! :balloon: :partying_face:

maximus19910112 commented:
bors merge

maximus19910112 commented:
bors merge

zsluedem commented:
bors delegate+

maximus19910112 commented:
bors r+
